File: without-library/graph/tree/dp_on_trees.py
import sys
sys.setrecursionlimit(200000)

class TreeDP:
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.dp = [[0, 0] for _ in range(n + 1)]
        
    def dfs(self, u, p):
        self.dp[u][0] = 0
        self.dp[u][1] = 1
        
        for v in self.adj[u]:
            if v != p:
                self.dfs(v, u)
                self.dp[u][0] += max(self.dp[v][0], self.dp[v][1])
                self.dp[u][1] += self.dp[v][0]

if __name__ == "__main__":
    pass

File: without-library/graph/tree/dp_on_trees.cpp
#include <iostream>
#include <vector>
#include <algorithm> // for max

using namespace std;

vector<int> adj[200005];
long long dp[200005][2];

void dfs_dp(int u, int p) {
    dp[u][0] = 0;
    dp[u][1] = 1;
    
    for(size_t i=0; i<adj[u].size(); i++) {
        int v = adj[u][i];
        if(v != p) {
            dfs_dp(v, u);
            dp[u][0] += max(dp[v][0], dp[v][1]);
            dp[u][1] += dp[v][0];
        }
    }
}

int main() {
    return 0;
}
